### PR TITLE
Add welcome-to-tech-palace exercise

### PR DESCRIPTION
**Pull request by ErikSchierboom**
Closes https://github.com/exercism/go/issues/1613

@junedev I haven't gotten around to adding the tests, but I can do that later. Could you review this PR?

### CODE REVIEW

File: concepts/strings-package/about.md
Comment (ErikSchierboom) [inline concepts/strings-package/about.md:7]: I have no idea what the most commonly used string functions are in Go. Please feel free to suggest any.

Comment (bobtfish) [reply concepts/strings-package/about.md:7]: `ToUpper`, `HasPrefix`, `HasSuffix`, `Replace`, `TrimSpace`, `Split`

Comment (ErikSchierboom) [reply concepts/strings-package/about.md:7]: Thanks

File: config.json
Comment (ErikSchierboom) [inline config.json:1852]: I had to add this as `logs-logs-logs` no longer has `strings-package` in its concepts and needed at least one concept.

**Comment by ErikSchierboom** [comment]
@junedev @ekingery This is ready for review. I don't know Go, so please check for any stupid mistakes :)

File: concepts/strings-package/.meta/config.json
Comment (iHiD) [inline concepts/strings-package/.meta/config.json:4]: I think you should be co-author here.

File: exercises/concept/welcome-to-tech-palace/.docs/instructions.md
Comment (iHiD) [inline exercises/concept/welcome-to-tech-palace/.docs/instructions.md:3]:
```suggestion
There is an appliance store called "Tech Palace" nearby. The owner of the store recently installed a big display to use for marketing messages and to show a special greetings when customers scan their loyalty cards at the entrance. The display consists of lots of small LED lights and can show multiple lines of text.
```

File: exercises/concept/welcome-to-tech-palace/.docs/instructions.md
Comment (iHiD) [inline exercises/concept/welcome-to-tech-palace/.docs/instructions.md:9]:
```suggestion
For most customers who scan their loyalty cards, the store owner wants to see `Welcome to the Tech Palace, ` following by the name of the customer in capital letters on the display.
```

"normal" has negative connotations to many people. I also added reference again to the loyalty card as I wondered inititally how the owner knew their names

File: exercises/concept/welcome-to-tech-palace/.docs/instructions.md
Comment (iHiD) [inline exercises/concept/welcome-to-tech-palace/.docs/instructions.md:20]:
```suggestion
For loyal customers that buy a lot at the store, the owner wants the welcome display to be more fancy by adding a line of stars before and after the welcome message. They are not sure yet how many stars should be in the lines so they want that to be configurable.
```

File: exercises/concept/welcome-to-tech-palace/.docs/instructions.md
Comment (iHiD) [inline exercises/concept/welcome-to-tech-palace/.docs/instructions.md:34]:
```suggestion
Before installing this new display, the store had a similar display that could only show non-configurable, static lines. The owner would like to reuse some of the old marketing messages  on the new display. However, the data already includes a star border and some unfortunate whitespaces. Your task is to clean up the messages so they can be re-used.
```

File: concepts/strings-package/.meta/config.json
Comment (ErikSchierboom) [reply concepts/strings-package/.meta/config.json]: Yeah make sense

File: concepts/strings/.meta/config.json
Comment (jmrunkle) [inline concepts/strings/.meta/config.json:2]: Pretty sure this blurb is wrong

File: concepts/strings/about.md
Comment (jmrunkle) [inline concepts/strings/about.md:3]: I am not certain if we want to specifically mention runes here. Perhaps not because you added another concept for `runes` that is supposed to be taught by `logs-logs-logs`.

Any preference @ekingery?

File: concepts/strings/introduction.md
Comment (jmrunkle) [inline concepts/strings/introduction.md:17]:
```suggestion
Some special characters need to be escaped with a leading backslash:
```

File: concepts/strings/links.json
Comment (jmrunkle) [inline concepts/strings/links.json:1]: https://go101.org/article/string.html is not too bad if you are looking for a decent reference to give

File: concepts/strings/introduction.md
Comment (jmrunkle) [inline concepts/strings/introduction.md:29]: Since there is another concept dedicated to this, should we put this here at all?

File: exercises/concept/welcome-to-tech-palace/.docs/instructions.md
Comment (jmrunkle) [inline exercises/concept/welcome-to-tech-palace/.docs/instructions.md:3]:
```suggestion
There is an appliance store called "Tech Palace" nearby.
The owner of the store recently installed a big display to use for marketing messages and to show a special greetings when customers scan their loyalty cards at the entrance.
The display consists of lots of small LED lights and can show multiple lines of text.
```

Mind breaking up the sentences one per line like this?

File: exercises/concept/welcome-to-tech-palace/.docs/instructions.md
Comment (jmrunkle) [inline exercises/concept/welcome-to-tech-palace/.docs/instructions.md:11]:
```suggestion
Implement the function `WelcomeMessage` that accepts the name of the customer as a string argument and returns the desired message as a string.
```

File: exercises/concept/welcome-to-tech-palace/.docs/instructions.md
Comment (jmrunkle) [inline exercises/concept/welcome-to-tech-palace/.docs/instructions.md:22]:
```suggestion
Write a function `AddBorder` that accepts a welcome message (a `string`) and the number of stars per line (type `int`) as arguments.
It should return a `string` that consists of 3 lines, a line with the desired number of stars, then the welcome message as it was passed in, then another line of stars.
```

We should probably be consistent if we are going to use backticks around types.

File: exercises/concept/welcome-to-tech-palace/.meta/design.md
Comment (jmrunkle) [inline exercises/concept/welcome-to-tech-palace/.meta/design.md:16]: Runes?

File: exercises/concept/welcome-to-tech-palace/welcome_to_tech_palace.go
Comment (jmrunkle) [inline exercises/concept/welcome-to-tech-palace/welcome_to_tech_palace.go:8]:
```suggestion
// AddBorder adds a border to a welcome message.
```

File: exercises/concept/welcome-to-tech-palace/welcome_to_tech_palace_test.go
Comment (jmrunkle) [inline exercises/concept/welcome-to-tech-palace/welcome_to_tech_palace_test.go:1]: I am not certain if this conforms to our current test guidance for the go track. The top-level README.md file suggests these should use t.Fatalf rather than t.Errorf. Perhaps that is out of date?

There also seems to be a mild preference for using `got` and `want` rather than `expected`.

@ekingery?

File: concepts/strings/about.md
Comment (ekingery) [reply concepts/strings/about.md:3]: Yes, as long as it is covered elsewhere, let's leave them out. In addition to the go101 link you recommend for the links.json below, I would defer to @ErikSchierboom as to whether this link is also appropriate: https://go.dev/blog/strings

It is much more comprehensive and does cover runes, which might be nice if people want to take a deeper dive with this concept.

File: exercises/concept/welcome-to-tech-palace/welcome_to_tech_palace_test.go
Comment (ekingery) [reply exercises/concept/welcome-to-tech-palace/welcome_to_tech_palace_test.go:1]: I'm pretty sure we do want `t.Fatalf`, so that the test runner stops after the first failure. Although I'm going from memory and I think @ErikSchierboom wrote the test runner spec / docs so I'll leave that up to him.

Yes, swapping out `expected` with `want` would make it more consistent with the docs and other tests across the track. Thanks!

**Review by ekingery — APPROVED** [review]
Looks great overall! I chimed in on a couple of comments, but didn't scrutinize the whole PR.

File: concepts/strings/about.md
Comment (junedev) [inline concepts/strings/about.md:27]: the backtick at the end seems obsolete

File: concepts/strings-package/about.md
Comment (junedev) [inline concepts/strings-package/about.md:9]: I think spaces means just "space characters", should say whitespace

File: concepts/strings/about.md
Comment (junedev) [inline concepts/strings/about.md:30]: Doesn't the single backtick create problems in the Markdown?

File: concepts/strings/about.md
Comment (junedev) [inline concepts/strings/about.md:35]: to be consistent with the example above there should be an empty line between the two names

File: exercises/concept/welcome-to-tech-palace/.docs/introduction.md
Comment (junedev) [inline exercises/concept/welcome-to-tech-palace/.docs/introduction.md:29]: @ErikSchierboom Without a list of more functions or a link to where to find more functions, the info in the introduction is currently not sufficient to solve the exercise.

File: exercises/concept/welcome-to-tech-palace/.meta/design.md
Comment (junedev) [inline exercises/concept/welcome-to-tech-palace/.meta/design.md:11]: the strings package should also be mentioned in the learning objectives

File: exercises/concept/welcome-to-tech-palace/go.mod
Comment (junedev) [inline exercises/concept/welcome-to-tech-palace/go.mod:1]: module name needs to be changed

File: concepts/strings/about.md
Comment (ErikSchierboom) [reply concepts/strings/about.md:35]: I update the example above to remove the extra newline. Well spotted.

File: concepts/strings-package/about.md
Comment (ErikSchierboom) [reply concepts/strings-package/about.md]: Totally

File: exercises/concept/welcome-to-tech-palace/.meta/design.md
Comment (ErikSchierboom) [reply exercises/concept/welcome-to-tech-palace/.meta/design.md:18]: Added

File: concepts/strings/about.md
Comment (ErikSchierboom) [reply concepts/strings/about.md]: Prettier didn't complain, but I added an escape character

File: exercises/concept/welcome-to-tech-palace/.docs/introduction.md
Comment (ErikSchierboom) [reply exercises/concept/welcome-to-tech-palace/.docs/introduction.md:29]: I disagree with this statement for the following reason. One of the core ideas of Concept Exercises was to have students figure things out by themselves as much as possible. The way I intended this to work is to provide _just_ enough context to allow the student to Google things by themselves, but to always provide the hints as a fallback should the student not manage to figure things out. Applying that to this example, we've explained that there is a `strings` package (letting the student learn its exact name, useful for Googling), shown how to import it and how to use one example function (purposefully _not_ one of the functions the student requires).

I know not everyone is comfortable with the above approach, but it _is_ the original idea for how to introduce concepts in Concept Exercises. If you as Go maintainers would prefer to use a different approach, I'm okay with that, but I think there's great value in the above approach.

File: exercises/concept/welcome-to-tech-palace/.docs/instructions.md
Comment (ErikSchierboom) [reply exercises/concept/welcome-to-tech-palace/.docs/instructions.md]: Agreed

File: concepts/strings/introduction.md
Comment (ErikSchierboom) [reply concepts/strings/introduction.md:29]: We shouldn't! Copy-paste error.

File: exercises/concept/welcome-to-tech-palace/.docs/instructions.md
Comment (ErikSchierboom) [reply exercises/concept/welcome-to-tech-palace/.docs/instructions.md]: 👍

File: exercises/concept/welcome-to-tech-palace/.docs/introduction.md
Comment (junedev) [reply exercises/concept/welcome-to-tech-palace/.docs/introduction.md:29]: I am ok either way, I was just a little confused because I had the feeling you were pushing a lot for providing all the needed info in other exercises. I got the wrong impression from that. Just ignore the comment.

File: exercises/concept/welcome-to-tech-palace/welcome_to_tech_palace_test.go
Comment (iHiD) [reply exercises/concept/welcome-to-tech-palace/welcome_to_tech_palace_test.go:1]: > I'm pretty sure we do want t.Fatalf, so that the test runner stops after the first failure. Although I'm going from memory and I think @ErikSchierboom wrote the test runner spec / docs so I'll leave that up to him.

I think all tests should run so we can list the passed/failed numbers to a student.

**Review by junedev — APPROVED** [review]

**Comment by junedev** [comment]
@ErikSchierboom Not sure whether this should be part of this PR or not but at some point we need to go through all exercises that currently have "string-formatting" as prerequisite and check whether they need that one or the new "strings". If you dont want to include this here maybe create a ticket so we dont forget about it.

File: exercises/concept/welcome-to-tech-palace/welcome_to_tech_palace_test.go
Comment (ErikSchierboom) [reply exercises/concept/welcome-to-tech-palace/welcome_to_tech_palace_test.go:1]: > I'm pretty sure we do want t.Fatalf, so that the test runner stops after the first failure. Although I'm going from memory and I think @ErikSchierboom wrote the test runner spec / docs so I'll leave that up to him.

We used to only display the first error, but have since reverted that decision:

image

> Yes, swapping out expected with want would make it more consistent with the docs and other tests across the track. Thanks!

Done. It might be worth opening an issue to convert other exercises too.

File: exercises/concept/welcome-to-tech-palace/.docs/introduction.md
Comment (ErikSchierboom) [reply exercises/concept/welcome-to-tech-palace/.docs/introduction.md:29]: No problem. I will get this wrong from time to time too, so 🤷

File: concepts/strings/links.json
Comment (ErikSchierboom) [reply concepts/strings/links.json:1]: @junedev What do you think about the above link? It does mention runes and slices.

**Comment by ErikSchierboom** [comment]
@junedev Excellent point. I added this commit: https://github.com/exercism/go/pull/1680/commits/f38828e6215c390a65f1bfc00d1865a9ffbbf066

**Comment by junedev** [comment]
> @junedev Excellent point. I added this commit: [f38828e](https://github.com/exercism/go/commit/f38828e6215c390a65f1bfc00d1865a9ffbbf066)

Thanks! Looks correct to me.

File: concepts/strings/links.json
Comment (junedev) [reply concepts/strings/links.json:1]: I think more advanced articles are ok for the links as this is labelled as "Read More" on the website.

Comment (ErikSchierboom) [reply concepts/strings/links.json:1]: 👍 I'll add it

**Comment by ErikSchierboom** [comment]
I'm merging this. Thanks all!

**Pull request merged**
